cmd/registry/patch/project.go

```diff
@@ -53,17 +52,56 @@ func ExportProject(ctx context.Context, client *gapic.RegistryClient, projectNam
 	}
 
 	return core.ListArtifacts(ctx, client, projectName.Artifact(""), "", false, func(message *rpc.Artifact) error {
-		bytes, header, err := ExportArtifact(ctx, client, message)
-		if err != nil {
-			log.FromContext(ctx).Warnf("Skipping %s: %s", message.Name, err)
-			return nil
-		}
-		if header.Kind == "Artifact" { // "Artifact" is the generic artifact type
-			log.FromContext(ctx).Warnf("Skipping %s", message.Name)
-			return nil
+		taskQueue <- &exportArtifactTask{
+			client:  client,
+			message: message,
+			dir:     artifactsDir,
 		}
-		log.FromContext(ctx).Infof("Exporting %s", message.Name)
-		filename := fmt.Sprintf("%s/%s.yaml", artifactsDir, header.Metadata.Name)
-		return os.WriteFile(filename, bytes, fs.ModePerm)
+		return nil
 	})
 }
+
+type exportAPITask struct {
+	client  connection.Client
+	message *rpc.Api
+	dir     string
+}
+
+func (task *exportAPITask) String() string {
+	return "export " + task.message.Name
+}
+
+func (task *exportAPITask) Run(ctx context.Context) error {
+	log.FromContext(ctx).Infof("Exporting %s", task.message.Name)
+	bytes, header, err := ExportAPI(ctx, task.client, task.message)
+	if err != nil {
+		return err
+	}
+	filename := fmt.Sprintf("%s/%s.yaml", task.dir, header.Metadata.Name)
+	return os.WriteFile(filename, bytes, 0644)
+}
+
+type exportArtifactTask struct {
+	client  connection.Client
+	message *rpc.Artifact
+	dir     string
+}
+
+func (task *exportArtifactTask) String() string {
+	return "export " + task.message.Name
+}
+
+func (task *exportArtifactTask) Run(ctx context.Context) error {
+	bytes, header, err := ExportArtifact(ctx, task.client, task.message)
+	if err != nil {
+		log.FromContext(ctx).Warnf("Skipping %s: %s", task.message.Name, err)
+		return nil
+	}
+	if header.Kind == "Artifact" { // "Artifact" is the generic artifact type
+		log.FromContext(ctx).Warnf("Skipping %s", task.message.Name)
+		return nil
+	}
+	log.FromContext(ctx).Infof("Exporting %s", task.message.Name)
```

Comment:
Should this be earlier in the function, before we attempt the export? If ExportArtifact produces any logs (either now or in the future) this will cause a strange log ordering.

Reply:
It's here because we might not export the artifact - the logic above decides if the artifact should be exported or not and logs "Skipping" messages when artifacts are skipped. To clarify this, I replaced the `-ing` endings with `-ed` here and in one other place for consistency.